File: backend/GameApp.Application/Services/GameUpdateService.cs
using GameApp.Application.UseCases;
using GameApp.Domain.Entities;
using GameApp.Domain.Repositories;
using GameApp.Domain.ValueObjects.Characters;
using GameApp.Domain.ValueObjects.Scenes;

namespace GameApp.Application.Services;

public class GameUpdateService : GameUpdateUseCase
{
    private readonly IGameRepository _repo;

    public GameUpdateService(IGameRepository repo) => _repo = repo;

    public async Task<Game?> UpdateGameAsync(Guid id, Character character, int numberScenesToFinish, List<Scene> completedScenes, NothingHappensScene finalScene)
    {
        var game = await _repo.FetchByIdAsync(id);
        if (game is null)
        {
            Console.WriteLine($"Game to update with id {id} not found."); 
            return null;
        }

        Game updatedGame = game.UpdateGame(character, numberScenesToFinish, completedScenes, finalScene);
        return await _repo.UpdateAsync(updatedGame);
    }
}

File: frontend/gameapp-frontend-js/src/application/services/game-services/game-update-service.js
import { IGameUpdateUseCase } from "../../usecases/game-use-cases/game-update-use-case";
import { IGameRepository } from "../../repositories/igame-repository";
import { Game } from "../../../domain/entities/game";

export class GameUpdateService extends IGameUpdateUseCase {
    constructor(gameRepository) {
        super();
        if (!(gameRepository instanceof IGameRepository)) {
            throw new TypeError("gameRepository must be an instance of IGameRepository");
        }
        this.gameRepository = gameRepository;
    }

    async updateGame(id, character, numberScenesToFinish, finalScene, listCurrentScenes, listCurrentUserActions, completedScenes, currentEnemy) {
        if (!id) throw new TypeError("id is required");

        const updatedGame = new Game(character, numberScenesToFinish, finalScene, listCurrentScenes, listCurrentUserActions, completedScenes, currentEnemy, id);

        await this.gameRepository.update(id, updatedGame);

        return updatedGame;
    }
}
